---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 
```{r}
library("UCSCXenaTools")
library("R.utils")
library("dplyr")
library("data.table")
library("readr")
#library("edgeR")
setwd("~/Desktop/rna_seq/TCGA")
source("~/workspace/bioinformatics_analysis/project_code/r_scripts/download_data.R")


```
Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 
~/workspace/bioinformatics_analysis/code/r_scripts/rna_seq_g6pd/TCGA_DESEQ.Rmd
```{r}
setwd("~/Desktop/rna_seq/TCGA")

data("XenaData")
write.csv(XenaData, "00_tblXenaHubInfo.csv")

## step 5-a
GeneExpectedCnt_toil = XenaGenerate(subset = XenaHostNames == "toilHub") %>% 
  XenaFilter(filterCohorts = "TCGA TARGET GTEx") %>%
  XenaFilter(filterDatasets = "TcgaTargetGtex_gene_expected_count");
XenaQuery(GeneExpectedCnt_toil) %>%
  XenaDownload(destdir = "./")
samples_gene_expression = fread("TcgaTargetGtex_gene_expected_count.gz");
```


```{r}
#g6pd_deleted_samples_file <-download_url("https://drive.google.com/file/d/1BxjcimPvrUOFOzLLbkwWc0fqyQGOLyju") #deleted samples
#sample_file_g6pd_control_normal_expression_file <- download_url("https://drive.google.com/file/d/1oRrh-GFLyv9J7lu2xChizBLKYMjHadu1") # 200 samples ctr
#sample_file_g6pd_control_normal_expression_file <- download_url("https://drive.google.com/file/d/18Io0Je293DopXAmz163wsALnf17UBxsu") # 2000 middle expression g6pd normal
#sample_file_g6pd_control_normal_expression_file <- download_url("https://drive.google.com/file/d/1fONWXejRx9xvKKJr-6cdbrhrMHK2PAvA") # propensity score controls for deleted samples
#sample_file_g6pd_control_normal_expression_file <- #download_url("https://drive.google.com/file/d/1VpjKXenSSWBewblp2GH7FWTV9n0ftwJB")#G6PD_normal_cancer_type_based_selection_not_metastatic.csv (~1400 samples)
TCGA_samples_data_path <- download_url("https://drive.google.com/file/d/1ELYxyn6lnF__Pc7YX2NtiitwWCCZNq-n", type="tsv")
#g6pd_deleted_samples_file <- download_url("https://drive.google.com/file/d/1eRDm68CTg2e4tUvJrJWWoYaJCPnNJm_v")#deleted and mutated samples together
#propensity_score_deleted_and_mutated_controls <- download_url("https://drive.google.com/file/d/139-ugep8A6yxC3MabSpDJN_nAqWP6A5C")
#propensity_score_deleted_and_mutated_treatment <-  download_url("https://drive.google.com/file/d/1fhugBDQBFy4DnbR_PswH8X_-KGgWxne2")

propensity_score_deleted_and_mutated_without_cancer_stage_matching <- download_url("https://drive.google.com/file/d/1wGqpQXIAfHiENczl7CAkSzmneIg-wtk3")
propensity_score_not_deleted_and_mutated_without_cancer_stage_matching <- download_url("https://drive.google.com/file/d/1sfvhe3iKx_NASMDFEcAweGtmnXrQD5MT")
propensity_score_deleted_and_mutated_without_cancer_stage_matching

propensity_score_not_deleted_and_mutated_without_cancer_stage_matching
```

```{r}
sample_file_g6pd_control_normal_expression_file <- propensity_score_not_deleted_and_mutated_without_cancer_stage_matching
g6pd_deleted_samples_file <- propensity_score_deleted_and_mutated_without_cancer_stage_matching
sample_file_g6pd_control_normal_expression_file
g6pd_deleted_samples_file
```


```{r}
TCGA_samples_data_df <- read_tsv(TCGA_samples_data_path)
read.csv(sample_file_g6pd_control_normal_expression_file)
read.csv(g6pd_deleted_samples_file)
```


```{r}
create_back_transformed_df <- function(sample_file, gene_expression_data) {
  # Read the sample list
  list_data <- read.csv(sample_file, header = FALSE)
  vector_data <- list_data$V1

  # Filter out excluded samples
  filtered_vector <- vector_data[vector_data %in% colnames(gene_expression_data)]


  # Filter the gene expression data
  filtered_df <- gene_expression_data %>% dplyr::select("sample", all_of(filtered_vector))
  rownames(filtered_df) <- filtered_df$sample
  
  # Backtransform the numeric columns
  backtransformed_df <- filtered_df %>%
    mutate(across(where(is.numeric), ~ 2^. - 1))
  rownames(backtransformed_df) <- backtransformed_df$sample
  # Set row names and remove the first column

  return(backtransformed_df)
}

#g6pd_deleted_samples_file <- "g6pd_deleted_samples.csv"
#sample_file_g6pd_control_normal_expression_file = "g6pd_normal_samples_control_group.csv"
#sample_file_g6pd_control_normal_expression_file = "g6pd_normal_control_200_samples.csv"

# Call the function and store the result
g6pd_del_df <- create_back_transformed_df(g6pd_deleted_samples_file, samples_gene_expression)
g6pd_normal_control_df <- create_back_transformed_df(sample_file_g6pd_control_normal_expression_file, samples_gene_expression)
#g6pd_normal_control_df
#g6pd_del_df

```


```{r}
# checking control data frame samples
# ctr_rnames <- colnames(g6pd_normal_control_df[,-1])
# ctr_orig_df <- read.csv("/Users/efraimshine/Desktop/rna_seq/TCGA/g6pd_high_exp_1.75to2.25_z_score_samples_names.csv")
# ctr_orig_df
# ctr_rnames
# common_names <- intersect(ctr_rnames,ctr_orig_df$Sample.ID)
# common_names
# print(common_names)
g6pd_normal_control_df
g6pd_del_df
```



```{r}
# Define the arguments
library("data.table")
library("readr")
library("dplyr")
#library("edgeR")




# Add prefixes to column names
#colnames(g6pd_del_df) <- paste0("DELE_", colnames(g6pd_del_df))
#colnames(g6pd_normal_control_df) <- paste0("NORM_", colnames(g6pd_normal_control_df))

# Combine the data frames by rows
merged_df <- merge(g6pd_del_df, g6pd_normal_control_df, by.x="sample",by.y = "sample")
#colnames(merged_df)[colnames(merged_df) == "DELE_sample"] <- "sample"
# rownames(merged_df) <- merged_df$sample
# merged_df$sample <- NULL

# rownames(g6pd_del_df) <- g6pd_del_df$DELE_sample
# g6pd_del_df$DELE_sample <- NULL

# rownames(g6pd_normal_control_df) <- g6pd_normal_control_df$NORM_sample
# g6pd_normal_control_df$NORM_sample <- NULL
head(merged_df)
first_36_names <- colnames(merged_df)[1:37]   # or names(df)[1:36]
colnames(g6pd_del_df)
first_36_names
setdiff(first_36_names, colnames(g6pd_del_df))


```


```{r}
merged_df_rownames <- merged_df[,-1]
rownames(merged_df_rownames) <- merged_df$sample
# merged_df_round_vector <- lapply(merged_df_rownames, round , digits = 2)
# merged_df_rownames_rounded <- data.frame(merged_df_round_vector)
#rownames(merged_df_rownames_rounded) <- rownames(merged_df_rownames)
merged_df_rownames_rounded <- merged_df_rownames %>% mutate(across(where(is.numeric), round, digits = 0))
matrix_for_deseq <- as.matrix(merged_df_rownames_rounded)
rownames(matrix_for_deseq) <- rownames(merged_df_rownames_rounded)
#merged_df_rownames
#merged_df_rownames_rounded
```

```{r}
#TCGA_samples_data_df
```


```{r}
library(dplyr)

wanted_TCGA_columns <- c("Sample ID", "Cancer Type", "Sample Type")

metadata <- data.frame(sample_names <- colnames(merged_df[,-1]))
#row.names(metadata) <- colnames(merged_df)
  metadata <- mutate(metadata, G6PD_STATUS = ifelse(row_number() <= length(g6pd_del_df[,-1]), "del", "normal"))
  
#metadata <- merge(metadata, TCGA_samples_data_df[, wanted_TCGA_columns], by.x = names(metadata)[1], by.y = "Sample ID", all.x = TRUE)

metadata <- metadata %>%
  left_join(TCGA_samples_data_df[, wanted_TCGA_columns], by = setNames("Sample ID", names(metadata)[1]))

names(metadata)[names(metadata) == "Cancer Type"] <- "cancer_type"

names(metadata)[names(metadata) == "Sample Type"] <- "sample_type"

metadata_rnames <- metadata
rownames(metadata_rnames) <- metadata$sample_names....colnames.merged_df
metadata_rnames$sample_names....colnames.merged_df....1.. <- NULL
#metadata
#metadata_rnames



```

```{r}
# metadata <- mutate(metadata, G6PD_STATUS = ifelse(row_number() <= length(g6pd_del_df[,-1]), "del", "normal"))

#setwd("/Users/efraimshine/Desktop/rna_seq/TCGA")
# metadata_rnames_for_info_merge <- metadata_rnames
#metadata_rnames_for_info_merge
#expanding meta data information for analysis of g6pd deleted samples VS 200 g6pd normal samples from the center of g6pd expression information
library("dplyr")

# samples_info_deleted_vs_200_patients <- read_tsv("g6pd_samples_info_deleted_vs_two_hundreds.tsv")
# samples_info_deleted_vs_200_patients <- as.data.frame(samples_info_deleted_vs_200_patients) %>% dplyr::select(`Sample ID`,`Cancer Type`,`Cancer Type Detailed`)
# 
# # merege metadata with info df
# row.names(metadata_rnames_for_info_merge) <- gsub("^[^_]*_", "", row.names(metadata_rnames_for_info_merge))
# 
# metadata_rnames_merged_with_info <- merge(metadata_rnames_for_info_merge,samples_info_deleted_vs_200_patients, by.x = "row.names", by.y = "Sample ID",all.x = TRUE)
# row.names(metadata_rnames_merged_with_info) <- metadata_rnames_merged_with_info$Row.names
# metadata_rnames_merged_with_info$Row.names <- NULL
# metadata_rnames_merged_with_info <- metadata_rnames_merged_with_info[order(metadata_rnames_merged_with_info$G6PD_STATUS), ]
# metadata_rnames_merged_with_info_orderd <- metadata_rnames_merged_with_info[match(row.names(metadata_rnames_for_info_merge), row.names(metadata_rnames_merged_with_info)), ]
# 
# row.names(metadata_rnames_merged_with_info_orderd) <- row.names(metadata_rnames)
# colnames(metadata_rnames_merged_with_info_orderd)[2] <- "cancer_type"
# colnames(metadata_rnames_merged_with_info_orderd)[3] <- "cancer_type_detailed"
# 
# metadata_rnames_merged_with_info_orderd$cancer_type <- as.factor(metadata_rnames_merged_with_info_orderd$cancer_type)
# metadata_rnames_merged_with_info_orderd$cancer_type_detailed <- as.factor(metadata_rnames_merged_with_info_orderd$cancer_type_detailed)
# metadata_rnames_merged_with_info
# metadata_rnames_merged_with_info_orderd

```




```{r}
names_df <- as.data.frame(data_col <- colnames(matrix_for_deseq),metadatarows <- rownames(metadata_rnames))
names_df$`data_col <- colnames(matrix_for_deseq)` == rownames(names_df)
#names_df
```



```{r}
library("DESeq2")
library("BiocParallel")
#rm(samples_gene_expression, XenaData); gc()
register(MulticoreParam(workers = 12)) 
dds <- DESeqDataSetFromMatrix(countData = matrix_for_deseq,
                              colData = metadata_rnames,
                              design = ~G6PD_STATUS)
dds$G6PD_STATUS <- relevel(dds$G6PD_STATUS, ref = "normal")
# smallestGroupSize <- 3
# keep <- rowSums(counts(dds) >= 10) >= smallestGroupSize
# dds <- dds[keep,]
dds
```
```{r}
dds <- DESeq(dds,parallel = TRUE)
res <- results(dds,parallel = TRUE)
res
```



```{r}
resultsNames(dds)

```
```{r}
resOrdered <- res[order(res$pvalue),]
summary(res)
```
```{r}
sum(res$padj < 0.05, na.rm=TRUE)
```
```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)
```
```{r}
plotMA(res, ylim=c(-2,2))
```
```{r}
# getting the transformed counts for pca
#rlog <- rlog(dds, blind = FALSE)
vsd <- vst(dds,blind = FALSE)
vsd
plotPCA(vsd, intgroup = "G6PD_STATUS")

```
```{r}
#explore pca results
pcaData <- plotPCA(vsd, intgroup = "G6PD_STATUS", returnData = TRUE)
pcaData 
cluster <- pcaData %>% filter(PC1 < -25)
cluster
```
```{r}
#convert ids to gene symbol 
library(org.Hs.eg.db)
ensembl_ids <- rownames(res)

ensembl_ids_clean <- sub("\\..*", "", ensembl_ids)
symbols <- mapIds(org.Hs.eg.db,
                  keys = ensembl_ids_clean,
                  keytype = "ENSEMBL",
                  column = "SYMBOL",
                  multiVals = "first")
head(symbols)
```
```{r}
# ── 1. Packages ──
library(AnnotationDbi)
library(org.Hs.eg.db)
library(pheatmap)



# ── 2. Gene set for GO:0009062 ──
go_id  <- "GO:0009062"
go2ens <- AnnotationDbi::select(org.Hs.eg.db,
                                keys     = go_id,
                                keytype  = "GOALL",
                                columns  = "ENSEMBL")
go_genes <- unique(go2ens$ENSEMBL)

# ── 3. Extract those genes from the VST matrix ──
vsd_mat        <- assay(vsd)
row_clean      <- sub("\\..*", "", rownames(vsd_mat))
keep_idx       <- which(row_clean %in% go_genes)
stopifnot(length(keep_idx) > 0)

vsd_subset <- vsd_mat[keep_idx, ]
vsd_subset <- t(scale(t(vsd_subset)))


# ── 4. Replace row names with symbols (optional) ──
gene_syms <- mapIds(org.Hs.eg.db, keys=row_clean[keep_idx],
                    keytype="ENSEMBL", column="SYMBOL", multiVals="first")
rownames(vsd_subset) <- ifelse(is.na(gene_syms), row_clean[keep_idx], gene_syms)

# ── 5. Column annotation ──
anno_df <- data.frame(G6PD_STATUS = metadata_rnames$G6PD_STATUS,
                      row.names   = rownames(metadata_rnames))
anno_df <- anno_df[colnames(vsd_subset), , drop=FALSE]

# ── 6. Order columns: cluster *inside* each G6PD group only ──
grp      <- anno_df$G6PD_STATUS
cols_norm <- which(grp == "normal")
cols_mut  <- which(grp != "normal")   # e.g. "del" / "mutated"

order_within <- function(idx) {
  if (length(idx) <= 1) return(idx)
  idx[hclust(dist(t(vsd_subset[, idx])), method = "average")$order]
}
ord_norm <- order_within(cols_norm)
ord_mut  <- order_within(cols_mut)

ord_cols <- c(ord_norm, ord_mut)      # normals first, mutants second
vsd_ord  <- vsd_subset[, ord_cols]
anno_ord <- anno_df[ord_cols, , drop=FALSE]

# ── 7. Draw the heat‑map ──
pheatmap <- pheatmap(vsd_ord,
         scale                     = "none",
         annotation_col            = anno_ord,
         cluster_cols              = FALSE,
         cluster_rows = TRUE , # we pre‑ordered columns
         gaps_col                  = length(ord_norm),  # visual break
         clustering_distance_rows  = "correlation",
         clustering_method         = "average",
         show_rownames             = TRUE,
         fontsize_row              = 6,
         main = "Fatty‑acid catabolic process (GO:0009062)")
pheatmap
# ── choose a perceptually uniform diverging palette ───────────────
# my_cols   <- colorRampPalette(c("#2166AC", "#F7F7F7", "#B2182B"))(100)  # blue‑white‑red
# my_breaks <- seq(-2, 2, length.out = 101)   # 0.04‑width bins
# 
# ## ── draw the heat‑map exactly as before but pass colour + breaks ──
# p <- pheatmap(vsd_ord,
#               scale                     = "row",
#               annotation_col            = anno_ord,
#               cluster_cols              = FALSE,
#               cluster_rows              = TRUE,
#               gaps_col                  = length(ord_norm),
#               clustering_distance_rows  = "correlation",
#               clustering_method         = "average",
#               color                     = my_cols,
#               breaks                    = my_breaks,
#               legend_breaks             = c(-2, -1, 0, 1, 2),
#               legend_labels             = c("‑2", "‑1", "0", "1", "2"),
#               show_rownames             = TRUE,
#               fontsize_row              = 6,
#               main = "Fatty‑acid catabolic process (GO:0009062)\n(row‑Z‑score, palette ±2 SD)")
# 
# p


```
```{r}
library(dplyr); library(stringr)
library(AnnotationDbi); library(org.Hs.eg.db); library(pheatmap)

## ── 1. Extract leading‑edge genes (ENSEMBL IDs) ────────────────────────────
go_id <- "GO:0009062"
lead_vec <- gsea_res@result %>% filter(ID == go_id) %>% pull(core_enrichment)
lead_ens <- str_split(lead_vec, "/")[[1]] %>% unique()

# optional: drop version numbers (e.g. ENSG00000123456.3 → ENSG00000123456)
lead_ens <- sub("\\..*$", "", lead_ens)

## ── 2. Build a matrix of mean‑centred log2FC instead of row‑scaling ───────
vsd_mat <- assay(vsd)                    # vst values (≈ log2 counts)
grp     <- metadata_rnames$G6PD_STATUS
stopifnot(identical(colnames(vsd_mat), rownames(metadata_rnames)))

# # Keep only leading‑edge genes that exist in the matrix
# row_clean <- sub("\\..*$", "", rownames(vsd_mat))
# keep_rows <- which(row_clean %in% lead_ens)
# if (length(keep_rows) == 0)
#     stop("No leading‑edge genes found in the vst matrix.")
# vsd_lead <- vsd_mat[keep_rows, ]

# Calculate group means and a per‑gene log2FC (del – normal)
mu_del    <- rowMeans(vsd_lead[, grp == "del",    drop = FALSE])
mu_normal <- rowMeans(vsd_lead[, grp == "normal", drop = FALSE])
vsd_centered <- vsd_lead - mu_normal              # centre on normals

## ── 3. Order columns: cluster within each group only ───────────────────────
order_within <- function(idx) {
  if (length(idx) <= 1) return(idx)
  idx[hclust(dist(t(vsd_centered[, idx])), method = "average")$order]
}
idx_norm <- which(grp == "normal")
idx_del  <- which(grp == "del")
ord_cols <- c(order_within(idx_norm), order_within(idx_del))

anno_df  <- data.frame(G6PD_STATUS = grp[ord_cols],
                       row.names   = colnames(vsd_centered)[ord_cols])

## ── 4. Replace row names with gene symbols for readability ────────────────
gene_syms <- mapIds(org.Hs.eg.db,
                    keys      = row_clean[keep_rows],
                    keytype   = "ENSEMBL",
                    column    = "SYMBOL",
                    multiVals = "first")
rownames(vsd_centered) <- ifelse(is.na(gene_syms),
                                 row_clean[keep_rows],
                                 gene_syms)

## ── 5. Draw the heat‑map ──────────────────────────────────────────────────
pheatmap(vsd_centered[, ord_cols],
         color                     = colorRampPalette(c("navy", "white", "firebrick3"))(100),
         annotation_col            = anno_df,
         cluster_cols              = FALSE,                 # we pre‑ordered
         gaps_col                  = length(idx_norm),      # visual split
         clustering_distance_rows  = "correlation",
         main = "Leading‑edge genes – fatty‑acid catabolic process (GO:0009062)\nvalues = expression centred on normal group")


```
```{r}

```

```{r}
# get expression plot for g6pd 
gene_of_interest <- "ENSG00000269087" #"ENSG00000160211"
# get gtpd id in the data set 
res_gene_ensble <- rownames(res)
matches <- grep(gene_of_interest, res_gene_ensble, value = TRUE)
print(matches)
# Get normalized counts
norm_counts <- counts(dds, normalized=TRUE)
# Select your gene


# Create a data frame
plot_df <- data.frame(
  expression = norm_counts[gene_of_interest, ],
  condition = colData(dds)$condition  # Replace 'condition' with your real grouping variable
)

library(ggplot2)

ggplot(plot_df, aes(x=condition, y=expression)) +
  geom_boxplot() +
  geom_jitter(width=0.2, color="blue", size=2) +
  theme_minimal() +
  labs(title=paste("Expression of", gene_of_interest),
       x="Condition",
       y="Normalized Expression")
```


```{r}
resSig <- subset(resOrdered, padj < 0.05)
resSig
# write.csv(as.data.frame(resSig), 
#           file="/Users/efraimshine/Desktop/rna_seq/TCGA/G6PD_abovepadj_0.05.csv")
```
```{r}
#filtering the results
#results_df <- read.csv("/Users/efraimshine/Desktop/rna_seq/TCGA/G6PD_abovepadj_0.05.csv")
results_df <- resSig %>% as.data.frame
results_filtered <- results_df %>% filter(log2FoldChange >0 ,padj < 0.05)
results_filtered
# write.csv(as.data.frame(results_filtered), 
#           file="/Users/efraimshine/Desktop/rna_seq/TCGA/deseq2_res_g6pd/G6PD_results_filtered.csv")
```
```{r}
# clustering
library(clusterProfiler)
library(org.Hs.eg.db)
library(AnnotationDbi)
```


```{r}
# results as in tutorial notebook
sigs <- na.omit(res)
sigs
sigs <- sigs[sigs$padj < 0.05 & sigs$baseMean > 50,]
sigs
genes_to_test <- rownames(sigs[sigs$log2FoldChange > 0.5,])

genes_to_test_no_dots <- gsub("\\..*", "", genes_to_test)


res_for_gsea <- res[res$baseMean > 50,] 
res_for_gsea <- res_for_gsea[order(-res_for_gsea$stat),]
res_for_gsea 

genes_for_gsea <- res_for_gsea$stat
names(genes_for_gsea) <- gsub("\\..*", "", rownames(res_for_gsea))





```
```{r}
#GO analysis
GO_results <- enrichGO(gene = genes_to_test_no_dots, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
path_df_enrich <- as.data.frame(GO_results)

GO_results %>% as.data.frame()
```


```{r}
gsea_res <- gseGO(genes_for_gsea,
             ont = "BP",
             keyType = "ENSEMBL",
             OrgDb = "org.Hs.eg.db",
             eps = 1e-300)
```
```{r}
gsea_res_df <- as.data.frame(gsea_res)
gsea_res_df
```
```{r}
fatty_acid_catabolic_process <- "GO:0009062"
atty_acid_beta_oxidation_using_acyl_CoA_dehydrogenase <-"GO:0033539"
lipid_catabolic_process	 <- "GO:0016042"
fit_gsea <- gseaplot(gsea_res, geneSetID = fatty_acid_catabolic_process ,title = "fatty acid catabolic process")
fit_gsea

```



```{r}
fit <- plot(barplot(GO_results, showCategory = 30))
fit
png("out.png", res = 250, width = 1400, height = 1800)
print(fit)
dev.off()

```


```{r}
# GO_results <- enricher()
# as.data.frame(GO_results)
```


```{r}
# prepaer meta data for multi factor analysis 
metadata_rnames_merged_with_info_orderd
metadata_rnames_merged_with_info_orderd$G6PD_STATUS <- factor(metadata_rnames_merged_with_info_orderd$G6PD_STATUS)
table(metadata_rnames_merged_with_info_orderd$cancer_type, metadata_rnames_merged_with_info_orderd$G6PD_STATUS)
metadata_rnames_merged_with_info_orderd$cancer_type <- droplevels(metadata_rnames_merged_with_info_orderd$cancer_type)
metadata_rnames_merged_with_info_orderd$G6PD_STATUS <- droplevels(metadata_rnames_merged_with_info_orderd$G6PD_STATUS)
metadata_rnames_merged_with_info_orderd
```



```{r}
# #multi factor analysis 
# metadata_rnames_merged_with_info_orderd
# library(DESeq2)
# ddsMF <- DESeqDataSetFromMatrix(countData = matrix_for_deseq,
#                               colData = metadata_rnames_merged_with_info_orderd,
#                               design = ~ cancer_type + G6PD_STATUS + cancer_type:G6PD_STATUS)
# ddsMF$G6PD_STATUS <- relevel(ddsMF$G6PD_STATUS, ref = "normal")
# ddsMF <- DESeq(ddsMF)
# #design(ddsMF) <- formula(~ cancer_type + G6PD_STATUS)
# # levels(ddsMF$`Cancer Type`)
# # levels(ddsMF$`Cancer Type`) <- sub("-.*", "", levels(ddsMF$`Cancer Type`))
# # levels(ddsMF$`Cancer Type`)s


```
```{r}
ddsMF_results <- results(ddsMF,contrast = c())
summary(ddsMF_results)
resultsNames(ddsMF_results)
```
```{r}
names(as.data.frame(coef(ddsMF)))
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.


